src/main.ml

```diff
@@ -83,68 +83,89 @@ let with_attached_vdi sr vdi read_write f =
+      | _, _, _ ->
+        fail (Failure "No suitable authentication provided")
+    ) >>= fun (session_id, need_to_logout) ->
+    Lwt.finalize
+      (fun () -> f session_id)
       (fun () ->
-        if need_to_logout
-        then Session.logout ~rpc ~session_id
-        else return ())
-  )
+         if need_to_logout
+         then Xen_api.Session.logout ~rpc ~session_id
+         else return ())
+  in
+
+  let channel = Nbd_lwt_unix.of_fd fd in
+  Lwt.finalize
+    (fun () ->
+       Nbd_lwt_unix.Server.connect channel ()
+       >>= fun (export_name, t) ->
+       Lwt.finalize
+         (fun () ->
+            let rpc = Xen_api.make !uri in
+            let uri = Uri.of_string export_name in
+            with_session rpc uri
+              (fun session_id ->
+                 let path = Uri.path uri in (* note preceeding / *)
+                 let vdi_uuid = if path <> "" then String.sub path 1 (String.length path - 1) else path in
+                 Xen_api.VDI.get_by_uuid ~rpc ~session_id ~uuid:vdi_uuid
+                 >>= fun vdi_ref ->
+                 Xen_api.VDI.get_record ~rpc ~session_id ~self:vdi_ref
+                 >>= fun vdi_rec ->
+                 Xen_api.SR.get_uuid ~rpc ~session_id ~self:vdi_rec.API.vDI_SR
+                 >>= fun sr_uuid ->
+                 with_attached_vdi sr_uuid vdi_rec.API.vDI_location (not vdi_rec.API.vDI_read_only)
+                   (fun filename ->
+                      with_block filename (Nbd_lwt_unix.Server.serve t (module Block))
+                   )
+              )
+         )
+         (fun () -> Nbd_lwt_unix.Server.close t)
+    )
+    (ignore_exn channel.close)
 
 let main port =
   let t =
     let sock = Lwt_unix.socket Unix.PF_INET Unix.SOCK_STREAM 0 in
-    Lwt_unix.setsockopt sock Lwt_unix.SO_REUSEADDR true;
-    let sockaddr = Lwt_unix.ADDR_INET(Unix.inet_addr_any, port) in
-    Lwt_unix.bind sock sockaddr;
-    Lwt_unix.listen sock 5;
-    let rec loop () =
-      begin
-        Lwt_unix.accept sock
-        >>= fun (fd, _) ->
-        (* Background thread per connection *)
-        let _ =
-          Lwt.catch
-            (fun () -> handle_connection fd)
-            (fun e -> Printf.fprintf stderr "Caught %s\n%!" (Printexc.to_string e); return ())
-          >>= fun () ->
-          Lwt_unix.close fd in
-        return ()
-      end >>= loop
-    in
-    loop ()
+    Lwt.finalize
+      (fun () ->
+         Lwt_unix.setsockopt sock Lwt_unix.SO_REUSEADDR true;
+         let sockaddr = Lwt_unix.ADDR_INET(Unix.inet_addr_any, port) in
+         Lwt_unix.bind sock sockaddr;
+         Lwt_unix.listen sock 5;
+         let rec loop () =
+           Lwt_unix.accept sock
+           >>= fun (fd, _) ->
+           (* Background thread per connection *)
+           let _ =
+             Lwt.catch
+               (fun () ->
+                  Lwt.finalize
+                    (fun () -> handle_connection fd)
+                    (ignore_exn (fun () -> Lwt_unix.close fd))
```

Comment:
I knew already, I just think it's better to stick it also in the code in this case